Lib/pyclbr.py

```diff
@@ -179,187 +172,121 @@ def _readmodule(module, path, inpackage=None):
     return _create_tree(fullmodule, path, fname, source, tree, inpackage)
 
 
-def _create_tree(fullmodule, path, fname, source, tree, inpackage):
-    """Return the tree for a particular module.
-
-    fullmodule (full module name), inpackage+module, becomes o.module.
-    path is passed to recursive calls of _readmodule.
-    fname becomes o.file.
-    source is tokenized.  Imports cause recursive calls to _readmodule.
-    tree is {} or {'__path__': <submodule search locations>}.
-    inpackage, None or string, is passed to recursive calls of _readmodule.
-
-    The effect of recursive calls is mutation of global _modules.
-    """
-    f = io.StringIO(source)
+class _ModuleBrowser(ast.NodeVisitor):
+    def __init__(self, module, path, file, tree, inpackage):
+        self.path = path
+        self.tree = tree
+        self.file = file
+        self.module = module
+        self.inpackage = inpackage
+        self.stack = []
+
+    def visit_ClassDef(self, node):
+        bases = []
+        for base in node.bases:
+            name = ast.unparse(base)
+            if name in self.tree:
+                # We know this super class.
+                bases.append(self.tree[name])
+            elif len(names := name.split(".")) > 1:
+                # Super class form is module.class:
+                # look in module for class.
+                *_, module, class_ = names
+                if module in _modules:
+                    bases.append(_modules[module].get(class_, name))
+            else:
+                bases.append(name)
+
+        parent = self.stack[-1] if self.stack else None
+        class_ = Class(
+            self.module, node.name, bases, self.file, node.lineno, parent
+        )
+        if parent is None:
+            self.tree[node.name] = class_
+        self.stack.append(class_)
+        self.generic_visit(node)
+        self.stack.pop()
+
+    def visit_Assign(self, node):
+        if not self.single_target_function_assign(node):
+            return
+
+        name = node.targets[0].id
+        value = self.tree[node.value.id]
+        parent = self.stack[-1]
+        child = Function(
+            value.module, name, value.file, node.lineno, parent, value.is_async
+        )
+        child.children = copy.deepcopy(value.children)
+        parent.children[name] = child
+        parent.methods[name] = node.lineno
+
+    def single_target_function_assign(self, node):
+        """Check if given assignment consists from a single target
+        and single value within a class namespace. Check value for if it
+        is an already defined function."""
+
+        return (len(node.targets) == 1
+                and len(self.stack) > 0
+                and isinstance(self.stack[-1], Class)
```

Comment:
```suggestion
                and isinstance(c:=self.stack[-1], Class)
```
See below for use of captured c.